Nightly failure: TestStdAlgorithmsTeamReduce error: no matching function for call to 'reduce'

**Describe the bug**
Nightly builds with clang/10.0 and c++20 (at least so far) began failing with output:

```
20:15:09 /home/jenkins/jenkins-new/workspace/Kokkos_develop_kokkos-dev-2_clang_100_cpp20-Wno-deprecated-volatile/kokkos/algorithms/unit_tests/TestStdAlgorithmsTeamReduce.cpp:207:34: error: no matching function for call to 'reduce'
20:15:09         const ValueType result = reduce(rowFromBegin, rowFromEnd);
20:15:09                                  ^~~~~~
20:15:09 /home/jenkins/jenkins-new/workspace/Kokkos_develop_kokkos-dev-2_clang_100_cpp20-Wno-deprecated-volatile/kokkos/algorithms/unit_tests/TestStdAlgorithmsTeamReduce.cpp:201:16: note: expanded from macro 'reduce'
20:15:09 #define reduce std::reduce
20:15:09                ^~~~~~~~~~~
20:15:09 /home/jenkins/jenkins-new/workspace/Kokkos_develop_kokkos-dev-2_clang_100_cpp20-Wno-deprecated-volatile/kokkos/algorithms/unit_tests/TestStdAlgorithmsTeamReduce.cpp:252:9: note: in instantiation of function template specialization 'Test::stdalgos::TeamReduce::test_A<Test::stdalgos::DynamicTag, double>' requested here
20:15:09         test_A<LayoutTag, ValueType>(numTeams, numCols, apiId);
20:15:09         ^
20:15:09 /home/jenkins/jenkins-new/workspace/Kokkos_develop_kokkos-dev-2_clang_100_cpp20-Wno-deprecated-volatile/kokkos/algorithms/unit_tests/TestStdAlgorithmsTeamReduce.cpp:259:3: note: in instantiation of function template specialization 'Test::stdalgos::TeamReduce::run_all_scenarios<Test::stdalgos::DynamicTag, double>' requested here
20:15:09   run_all_scenarios<DynamicTag, double>();
20:15:09   ^
20:15:09 /projects/sems/install/rhel7-x86_64/sems/compiler/gcc/9.2.0/base/lib/gcc/x86_64-pc-linux-gnu/9.2.0/../../../../include/c++/9.2.0/pstl/glue_numeric_defs.h:31:1: note: candidate function template not viable: requires 3 arguments, but 2 were provided
20:15:09 reduce(_ExecutionPolicy&& __exec, _ForwardIterator __first, _ForwardIterator __last);
20:15:09 ^
20:15:09 /projects/sems/install/rhel7-x86_64/sems/compiler/gcc/9.2.0/base/lib/gcc/x86_64-pc-linux-gnu/9.2.0/../../../../include/c++/9.2.0/pstl/glue_numeric_defs.h:26:1: note: candidate function template not viable: requires 4 arguments, but 2 were provided
20:15:09 reduce(_ExecutionPolicy&& __exec, _ForwardIterator __first, _ForwardIterator __last, _Tp __init);
20:15:09 ^
20:15:09 /projects/sems/install/rhel7-x86_64/sems/compiler/gcc/9.2.0/base/lib/gcc/x86_64-pc-linux-gnu/9.2.0/../../../../include/c++/9.2.0/pstl/glue_numeric_defs.h:21:1: note: candidate function template not viable: requires 5 arguments, but 2 were provided
20:15:09 reduce(_ExecutionPolicy&& __exec, _ForwardIterator __first, _ForwardIterator __last, _Tp __init,
```

This began following merge of the following commits:

**Changes**

implementation and tests ([details](https://jenkins-son.sandia.gov/job/Kokkos_develop_kokkos-dev-2_clang_100_cpp20-Wno-deprecated-volatile/961/changes#61839dccfcc7a67f475251bd9cba47b4f105da09) / [githubweb](https://github.com/kokkos/kokkos/commit/61839dccfcc7a67f475251bd9cba47b4f105da09))
    fix lambda capture and constraints ([details](https://jenkins-son.sandia.gov/job/Kokkos_develop_kokkos-dev-2_clang_100_cpp20-Wno-deprecated-volatile/961/changes#5742e3f71ef6859cbd42aa0d437b43d93bb78a6e) / [githubweb](https://github.com/kokkos/kokkos/commit/5742e3f71ef6859cbd42aa0d437b43d93bb78a6e))
    keep only subset ([details](https://jenkins-son.sandia.gov/job/Kokkos_develop_kokkos-dev-2_clang_100_cpp20-Wno-deprecated-volatile/961/changes#a3a00a6a2cda88201805cb4c20844f25265954bd) / [githubweb](https://github.com/kokkos/kokkos/commit/a3a00a6a2cda88201805cb4c20844f25265954bd))
    revert files ([details](https://jenkins-son.sandia.gov/job/Kokkos_develop_kokkos-dev-2_clang_100_cpp20-Wno-deprecated-volatile/961/changes#69389fcd35574260324713295aea0485f5f428ef) / [githubweb](https://github.com/kokkos/kokkos/commit/69389fcd35574260324713295aea0485f5f428ef))
    remove file ([details](https://jenkins-son.sandia.gov/job/Kokkos_develop_kokkos-dev-2_clang_100_cpp20-Wno-deprecated-volatile/961/changes#c46aa09065d5aaf4c3251cbaef37a21a6c2699e8) / [githubweb](https://github.com/kokkos/kokkos/commit/c46aa09065d5aaf4c3251cbaef37a21a6c2699e8))
    fix syntax ([details](https://jenkins-son.sandia.gov/job/Kokkos_develop_kokkos-dev-2_clang_100_cpp20-Wno-deprecated-volatile/961/changes#f5dde3d38d9108456e3961849497ce627f1eaaca) / [githubweb](https://github.com/kokkos/kokkos/commit/f5dde3d38d9108456e3961849497ce627f1eaaca))
    format ([details](https://jenkins-son.sandia.gov/job/Kokkos_develop_kokkos-dev-2_clang_100_cpp20-Wno-deprecated-volatile/961/changes#cb4ac66e83e83eccdfd519521ea3cf3827b57c25) / [githubweb](https://github.com/kokkos/kokkos/commit/cb4ac66e83e83eccdfd519521ea3cf3827b57c25))
    address comments ([details](https://jenkins-son.sandia.gov/job/Kokkos_develop_kokkos-dev-2_clang_100_cpp20-Wno-deprecated-volatile/961/changes#afac7784f662550c9c7b6d2fb4be2ac19bd41a0a) / [githubweb](https://github.com/kokkos/kokkos/commit/afac7784f662550c9c7b6d2fb4be2ac19bd41a0a))
    formatting ([details](https://jenkins-son.sandia.gov/job/Kokkos_develop_kokkos-dev-2_clang_100_cpp20-Wno-deprecated-volatile/961/changes#db591e3698fe5e8a440e19db0e8ec3fefa0bef24) / [githubweb](https://github.com/kokkos/kokkos/commit/db591e3698fe5e8a440e19db0e8ec3fefa0bef24))
    formatting ([details](https://jenkins-son.sandia.gov/job/Kokkos_develop_kokkos-dev-2_clang_100_cpp20-Wno-deprecated-volatile/961/changes#7d9394ddb770d0257229486a333ec020e1e78394) / [githubweb](https://github.com/kokkos/kokkos/commit/7d9394ddb770d0257229486a333ec020e1e78394))
    improve tests with intra-team result check ([details](https://jenkins-son.sandia.gov/job/Kokkos_develop_kokkos-dev-2_clang_100_cpp20-Wno-deprecated-volatile/961/changes#541f674684182b35e7498fbb38ccb96fffc15d0b) / [githubweb](https://github.com/kokkos/kokkos/commit/541f674684182b35e7498fbb38ccb96fffc15d0b))
    try fix for unique, previous impl to remove later ([details](https://jenkins-son.sandia.gov/job/Kokkos_develop_kokkos-dev-2_clang_100_cpp20-Wno-deprecated-volatile/961/changes#8bff82ff8056e4049991f0fd9a3136a8a6980293) / [githubweb](https://github.com/kokkos/kokkos/commit/8bff82ff8056e4049991f0fd9a3136a8a6980293))
    #5635: Add parallel_scan changes for CUDA and TeamThreadRange ([details](https://jenkins-son.sandia.gov/job/Kokkos_develop_kokkos-dev-2_clang_100_cpp20-Wno-deprecated-volatile/961/changes#1fb6f4a747dded20819f3894f1e51d1d1b8c8006) / [githubweb](https://github.com/kokkos/kokkos/commit/1fb6f4a747dded20819f3894f1e51d1d1b8c8006))
    remove old impl ([details](https://jenkins-son.sandia.gov/job/Kokkos_develop_kokkos-dev-2_clang_100_cpp20-Wno-deprecated-volatile/961/changes#db849838913c2ee9900d91db362bb4162dd83ad4) / [githubweb](https://github.com/kokkos/kokkos/commit/db849838913c2ee9900d91db362bb4162dd83ad4))
    #5635: Enable TeamThreadRange test for CUDA ([details](https://jenkins-son.sandia.gov/job/Kokkos_develop_kokkos-dev-2_clang_100_cpp20-Wno-deprecated-volatile/961/changes#6d794df9923b1a5f8ee548f79cdf8916dfd72cfb) / [githubweb](https://github.com/kokkos/kokkos/commit/6d794df9923b1a5f8ee548f79cdf8916dfd72cfb))
    OpenMPTarget: Disable check for SIMD compiler macros ([details](https://jenkins-son.sandia.gov/job/Kokkos_develop_kokkos-dev-2_clang_100_cpp20-Wno-deprecated-volatile/961/changes#7284cd215b3cd48caaecb2d22c73a082521b6622) / [githubweb](https://github.com/kokkos/kokkos/commit/7284cd215b3cd48caaecb2d22c73a082521b6622))
    add missing assert ([details](https://jenkins-son.sandia.gov/job/Kokkos_develop_kokkos-dev-2_clang_100_cpp20-Wno-deprecated-volatile/961/changes#2b7eb0b0b28a02e8bba3f729cc8d96a3cddf8c3f) / [githubweb](https://github.com/kokkos/kokkos/commit/2b7eb0b0b28a02e8bba3f729cc8d96a3cddf8c3f))
    add intra team check for missing test ([details](https://jenkins-son.sandia.gov/job/Kokkos_develop_kokkos-dev-2_clang_100_cpp20-Wno-deprecated-volatile/961/changes#ba9165994a95d6e0f168e70e5662d6bd52b8ce46) / [githubweb](https://github.com/kokkos/kokkos/commit/ba9165994a95d6e0f168e70e5662d6bd52b8ce46))
    fix intel compile error ([details](https://jenkins-son.sandia.gov/job/Kokkos_develop_kokkos-dev-2_clang_100_cpp20-Wno-deprecated-volatile/961/changes#bcb92a619e39a473a141f8ca8b023ba96ac68406) / [githubweb](https://github.com/kokkos/kokkos/commit/bcb92a619e39a473a141f8ca8b023ba96ac68406))
    fix unreachable for intel ([details](https://jenkins-son.sandia.gov/job/Kokkos_develop_kokkos-dev-2_clang_100_cpp20-Wno-deprecated-volatile/961/changes#e950759306bf5bd3500ba91a2f8d044240dbd9d8) / [githubweb](https://github.com/kokkos/kokkos/commit/e950759306bf5bd3500ba91a2f8d044240dbd9d8))
    re-enable unit tests for sort and random via makefile (#6422) ([details](https://jenkins-son.sandia.gov/job/Kokkos_develop_kokkos-dev-2_clang_100_cpp20-Wno-deprecated-volatile/961/changes#56cc35bdffcbbb7499070f3982779a297e705e1a) / [githubweb](https://github.com/kokkos/kokkos/commit/56cc35bdffcbbb7499070f3982779a297e705e1a))
    OpenMPTarget init-join fix (#6444) ([details](https://jenkins-son.sandia.gov/job/Kokkos_develop_kokkos-dev-2_clang_100_cpp20-Wno-deprecated-volatile/961/changes#7e274963234151ebd41ed9ee3d387ebece1a332e) / [githubweb](https://github.com/kokkos/kokkos/commit/7e274963234151ebd41ed9ee3d387ebece1a332e))
    Fix compiling code using Kokkos::printf for OpenMPTarget on Intel GPUs (#6443) ([details](https://jenkins-son.sandia.gov/job/Kokkos_develop_kokkos-dev-2_clang_100_cpp20-Wno-deprecated-volatile/961/changes#8239de526dcd028cd981e1db1cc731f5b078a15c) / [githubweb](https://github.com/kokkos/kokkos/commit/8239de526dcd028cd981e1db1cc731f5b078a15c))
    std_algos: improving min, max, minmax (#6421) ([details](https://jenkins-son.sandia.gov/job/Kokkos_develop_kokkos-dev-2_clang_100_cpp20-Wno-deprecated-volatile/961/changes#d8fa85644021d44452928cbd27c95ce992d5a538) / [githubweb](https://github.com/kokkos/kokkos/commit/d8fa85644021d44452928cbd27c95ce992d5a538))
    team-level stdalgos: improve tests, check intra-team result matching (part 2/7) (#6426) ([details](https://jenkins-son.sandia.gov/job/Kokkos_develop_kokkos-dev-2_clang_100_cpp20-Wno-deprecated-volatile/961/changes#3cd2813765a715e8730e454ae64996733ac0a5ca) / [githubweb](https://github.com/kokkos/kokkos/commit/3cd2813765a715e8730e454ae64996733ac0a5ca))
    Skip bessel function tests known to fail on Intel GPUs (#6434) ([details](https://jenkins-son.sandia.gov/job/Kokkos_develop_kokkos-dev-2_clang_100_cpp20-Wno-deprecated-volatile/961/changes#5ed274c9370896870637054a35b88373f1fd147f) / [githubweb](https://github.com/kokkos/kokkos/commit/5ed274c9370896870637054a35b88373f1fd147f))
    team-level stdalgos: improve tests, check intra-team result matching (part 6/7) (#6436) ([details](https://jenkins-son.sandia.gov/job/Kokkos_develop_kokkos-dev-2_clang_100_cpp20-Wno-deprecated-volatile/961/changes#e13f67ca64a83ab5ff85cf9937df179ca176736f) / [githubweb](https://github.com/kokkos/kokkos/commit/e13f67ca64a83ab5ff85cf9937df179ca176736f))


**Please also include the following items to support reproducing the bug**
**Reproducer** (kokkos-dev-2):
```bash
  module purge
  source /projects/sems/modulefiles/utils/sems-archive-modules-init.sh
  module load sems-archive-env sems-archive-cmake/3.17.1 sems-archive-gcc/9.2.0 sems-archive-clang/10.0.0

$KOKKOS_PATH/generate_makefile.bash --with-devices=Serial --arch=Volta70 --compiler=clang++ --cxxflags="-O3 -Wall -Wunused-parameter -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wuninitialized -Wno-deprecated-volatile" --cxxstandard="20" --kokkos-path=$KOKKOS_PATH --no-examples   --with-options=enable_large_mem_tests

```

Thanks for the work on this @fnrizzi !

To update the issue, there were also nightly failures with Sycl on Intel GPUs with intel-oneapi/2023.1.0:
```
01:38:55 /home/jenkins/blake-new/workspace/Kokkos_develop_blake_sycl_pv_oneapi_2023_1_0/kokkos/algorithms/unit_tests/TestStdAlgorithmsTeamReduce.cpp:207:34: error: no member named 'reduce' in namespace 'std'; did you mean 'sycl::ext::oneapi::reduce'?
01:38:55         const ValueType result = reduce(rowFromBegin, rowFromEnd);
01:38:55                                  ^~~~~~
01:38:55                                  sycl::ext::oneapi::reduce
01:38:55 /home/jenkins/blake-new/workspace/Kokkos_develop_blake_sycl_pv_oneapi_2023_1_0/kokkos/algorithms/unit_tests/TestStdAlgorithmsTeamReduce.cpp:201:16: note: expanded from macro 'reduce'
01:38:55 #define reduce std::reduce
01:38:55                ^~~~~
01:38:55 /projects/x86-64-icelake-rocky8/compilers/intel-oneapi-compilers/2023.1.0/gcc/8.5.0/base/6g2jkiv/compiler/2023.1.0/linux/bin-llvm/../include/sycl/ext/oneapi/group_algorithm.hpp:270:24: note: 'sycl::ext::oneapi::reduce' declared here
01:38:55                     T> reduce(Group g, T x, BinaryOperation binary_op) {
```

For reference, here are the reproducer instructions for the Blake testbed:
```bash
# Interactive node with PonteVecchio GPUs
salloc -N 1 -p PV

# Environment
module purge
module load cmake intel-oneapi-compilers/2023.1.0 intel-oneapi-dpl/2022.1.0 git

# Configuration
$KOKKOS_PATH/generate_makefile.bash --with-sycl --arch=INTEL_PVC --compiler=/projects/x86-64-icelake-rocky8/compilers/intel-oneapi-compilers/2023.1.0/gcc/8.5.0/base/6g2jkiv/compiler/2023.1.0/linux/bin-llvm/clang++ --cxxflags="-fp-model=precise" --cmake-flags=-DKokkos_ENABLE_ONEDPL=OFF
```

If you think #6459 should cover this case I can test it out. Compilation times are pretty long for this build, just to let you know

Also failures with rocm/5.2.0 , gcc/8.5.0 is loaded as part of the toolchain so hopefully https://github.com/kokkos/kokkos/pull/6459 resolves that, I can test that on caraway

**Error snip**:
```
22:35:53 /home/jenkins/caraway-new/workspace/Kokkos_develop_carawaymi50_rocm520_hipserial/kokkos/algorithms/unit_tests/TestStdAlgorithmsTeamReduce.cpp:207:34: error: no member named 'reduce' in namespace 'std'; did you mean 'Kokkos::Experimental::reduce'?
22:35:53         const ValueType result = reduce(rowFromBegin, rowFromEnd);
22:35:53                                  ^~~~~~
22:35:53                                  Kokkos::Experimental::reduce
```

**Reproducer instructions** for reference (Caraway testbed):
```bash
# Interactive node on MI50 queue
salloc -N 1 -p MI50

# Environment
module purge
module load cmake/3.19.3 rocm/5.2.0

# Configuration
$KOKKOS_PATH/generate_makefile.bash --with-devices=Hip,Serial --arch=VEGA906 --compiler=hipcc --cxxflags="-O3 -Werror " --cxxstandard="17" --kokkos-path=$KOKKOS_PATH --no-examples   --with-options=enable_large_mem_tests

```

--- a/algorithms/unit_tests/TestStdAlgorithmsCommon.hpp
+++ b/algorithms/unit_tests/TestStdAlgorithmsCommon.hpp
@@ -254,7 +254,7 @@ auto make_bounds(const ValueType1& lower, const ValueType2 upper) {
   return Kokkos::pair<ValueType1, ValueType2>{lower, upper};
 }
 
-#if defined(__GNUC__) && (KOKKOS_COMPILER_GNU <= 920)
+#if defined(_GLIBCXX_RELEASE) && (_GLIBCXX_RELEASE <= 9)
 
 // GCC 8 does not have reduce, transform_reduce, exclusive_scan, inclusive_scan,
 //       transform_exclusive_scan, transform_inclusive_scan

--- a/algorithms/unit_tests/TestStdAlgorithmsCommon.hpp
+++ b/algorithms/unit_tests/TestStdAlgorithmsCommon.hpp
@@ -254,12 +254,12 @@ auto make_bounds(const ValueType1& lower, const ValueType2 upper) {
   return Kokkos::pair<ValueType1, ValueType2>{lower, upper};
 }
 
+// libstdc++ as provided by GCC 8 does not have reduce, transform_reduce,
+// exclusive_scan, inclusive_scan, transform_exclusive_scan,
+// transform_inclusive_scan and for GCC 9.1, 9.2 fails to compile them for
+// missing overload not accepting policy so use here simplified versions of
+// them, only for testing purpose
 #if defined(_GLIBCXX_RELEASE) && (_GLIBCXX_RELEASE <= 9)
-
-// GCC 8 does not have reduce, transform_reduce, exclusive_scan, inclusive_scan,
-//       transform_exclusive_scan, transform_inclusive_scan
-// GCC 9.1, 9.2 fails to compile them for missing overload not accepting policy
-// so use here simplified versions of them, only for testing purpose
 
 template <class InputIterator, class ValueType, class BinaryOp>
 ValueType testing_reduce(InputIterator first, InputIterator last,

--- a/algorithms/unit_tests/TestStdAlgorithmsCommon.hpp
+++ b/algorithms/unit_tests/TestStdAlgorithmsCommon.hpp
@@ -254,11 +254,12 @@ auto make_bounds(const ValueType1& lower, const ValueType2 upper) {
   return Kokkos::pair<ValueType1, ValueType2>{lower, upper};
 }
 
-#if defined(__GNUC__) && __GNUC__ == 8
+#if defined(__GNUC__) && (KOKKOS_COMPILER_GNU <= 920)
 
-// GCC 8 doesn't come with reduce, transform_reduce, exclusive_scan,
-// inclusive_scan, transform_exclusive_scan and transform_inclusive_scan so here
-// are simplified versions of them, only for testing purpose
+// GCC 8 does not have reduce, transform_reduce, exclusive_scan, inclusive_scan,
+//       transform_exclusive_scan, transform_inclusive_scan
+// GCC 9.1, 9.2 fails to compile them for missing overload not accepting policy
+// so use here simplified versions of them, only for testing purpose
 
 template <class InputIterator, class ValueType, class BinaryOp>
 ValueType testing_reduce(InputIterator first, InputIterator last,

--- a/algorithms/unit_tests/TestStdAlgorithmsTeamReduce.cpp
+++ b/algorithms/unit_tests/TestStdAlgorithmsTeamReduce.cpp
@@ -196,7 +196,7 @@ void test_A(std::size_t numTeams, std::size_t numCols, int apiId) {
 
 // GCC 8 does not have reduce
 // GCC 9.1, 9.2 fails to compile for missing overload not accepting policy
-#if defined(__GNUC__) && (KOKKOS_COMPILER_GNU <= 920)
+#if defined(_GLIBCXX_RELEASE) && (_GLIBCXX_RELEASE <= 9)
 #define reduce testing_reduce
 #else
 #define reduce std::reduce

--- a/algorithms/unit_tests/TestStdAlgorithmsTeamReduce.cpp
+++ b/algorithms/unit_tests/TestStdAlgorithmsTeamReduce.cpp
@@ -197,8 +197,7 @@ void test_A(std::size_t numTeams, std::size_t numCols, int apiId) {
     // libstdc++ as provided by GCC 8 does not have reduce, transform_reduce,
     // exclusive_scan, inclusive_scan, transform_exclusive_scan,
     // transform_inclusive_scan and for GCC 9.1, 9.2 fails to compile them for
-    // missing overload not accepting policy so use here simplified versions of
-    // them, only for testing purpose
+    // missing overload not accepting policy
 
 #if defined(_GLIBCXX_RELEASE) && (_GLIBCXX_RELEASE <= 9)
 #define reduce testing_reduce

--- a/algorithms/unit_tests/TestStdAlgorithmsTeamReduce.cpp
+++ b/algorithms/unit_tests/TestStdAlgorithmsTeamReduce.cpp
@@ -194,8 +194,12 @@ void test_A(std::size_t numTeams, std::size_t numCols, int apiId) {
 
     ASSERT_TRUE(intraTeamSentinelView_h(i));
 
-// GCC 8 does not have reduce
-// GCC 9.1, 9.2 fails to compile for missing overload not accepting policy
+    // libstdc++ as provided by GCC 8 does not have reduce, transform_reduce,
+    // exclusive_scan, inclusive_scan, transform_exclusive_scan,
+    // transform_inclusive_scan and for GCC 9.1, 9.2 fails to compile them for
+    // missing overload not accepting policy so use here simplified versions of
+    // them, only for testing purpose
+
 #if defined(_GLIBCXX_RELEASE) && (_GLIBCXX_RELEASE <= 9)
 #define reduce testing_reduce
 #else

--- a/algorithms/unit_tests/TestStdAlgorithmsTeamReduce.cpp
+++ b/algorithms/unit_tests/TestStdAlgorithmsTeamReduce.cpp
@@ -194,8 +194,9 @@ void test_A(std::size_t numTeams, std::size_t numCols, int apiId) {
 
     ASSERT_TRUE(intraTeamSentinelView_h(i));
 
-// GCC 8 does not have reduce so guard against this
-#if defined(__GNUC__) && __GNUC__ == 8
+// GCC 8 does not have reduce
+// GCC 9.1, 9.2 fails to compile for missing overload not accepting policy
+#if defined(__GNUC__) && (KOKKOS_COMPILER_GNU <= 920)
 #define reduce testing_reduce
 #else
 #define reduce std::reduce

--- a/algorithms/unit_tests/TestStdAlgorithmsTeamTransformReduce.cpp
+++ b/algorithms/unit_tests/TestStdAlgorithmsTeamTransformReduce.cpp
@@ -242,8 +242,7 @@ void test_A(std::size_t numTeams, std::size_t numCols, int apiId) {
 // libstdc++ as provided by GCC 8 does not have reduce, transform_reduce,
 // exclusive_scan, inclusive_scan, transform_exclusive_scan,
 // transform_inclusive_scan and for GCC 9.1, 9.2 fails to compile them for
-// missing overload not accepting policy so use here simplified versions of
-// them, only for testing purpose
+// missing overload not accepting policy
 #if defined(_GLIBCXX_RELEASE) && (_GLIBCXX_RELEASE <= 9)
 #define transform_reduce testing_transform_reduce
 #else

--- a/algorithms/unit_tests/TestStdAlgorithmsTeamTransformReduce.cpp
+++ b/algorithms/unit_tests/TestStdAlgorithmsTeamTransformReduce.cpp
@@ -241,7 +241,7 @@ void test_A(std::size_t numTeams, std::size_t numCols, int apiId) {
 
 // GCC 8 does not have transform_reduce
 // GCC 9.1, 9.2 fails to compile for missing overload not accepting policy
-#if defined(__GNUC__) && (KOKKOS_COMPILER_GNU <= 920)
+#if defined(_GLIBCXX_RELEASE) && (_GLIBCXX_RELEASE <= 9)
 #define transform_reduce testing_transform_reduce
 #else
 #define transform_reduce std::transform_reduce

--- a/algorithms/unit_tests/TestStdAlgorithmsTeamTransformReduce.cpp
+++ b/algorithms/unit_tests/TestStdAlgorithmsTeamTransformReduce.cpp
@@ -239,8 +239,11 @@ void test_A(std::size_t numTeams, std::size_t numCols, int apiId) {
 
     ASSERT_TRUE(intraTeamSentinelView_h(i));
 
-// GCC 8 does not have transform_reduce
-// GCC 9.1, 9.2 fails to compile for missing overload not accepting policy
+// libstdc++ as provided by GCC 8 does not have reduce, transform_reduce,
+// exclusive_scan, inclusive_scan, transform_exclusive_scan,
+// transform_inclusive_scan and for GCC 9.1, 9.2 fails to compile them for
+// missing overload not accepting policy so use here simplified versions of
+// them, only for testing purpose
 #if defined(_GLIBCXX_RELEASE) && (_GLIBCXX_RELEASE <= 9)
 #define transform_reduce testing_transform_reduce
 #else

--- a/algorithms/unit_tests/TestStdAlgorithmsTeamTransformReduce.cpp
+++ b/algorithms/unit_tests/TestStdAlgorithmsTeamTransformReduce.cpp
@@ -239,8 +239,9 @@ void test_A(std::size_t numTeams, std::size_t numCols, int apiId) {
 
     ASSERT_TRUE(intraTeamSentinelView_h(i));
 
-// GCC 8 does not have transform_reduce so guard against this
-#if defined(__GNUC__) && __GNUC__ == 8
+// GCC 8 does not have transform_reduce
+// GCC 9.1, 9.2 fails to compile for missing overload not accepting policy
+#if defined(__GNUC__) && (KOKKOS_COMPILER_GNU <= 920)
 #define transform_reduce testing_transform_reduce
 #else
 #define transform_reduce std::transform_reduce

